wapitiCore/attack/mod_wapp.py

```diff
@@ -440,20 +444,32 @@ async def _detect_applications_headless(self, url: str) -> dict:
                         try:
                             results = await headless_client.execute_script(script)
                             for software, version_and_js in results.items():
-                                version, js = version_and_js
-                                expected_format = data[software]["js"][js]
-                                if version == "undefined":
-                                    continue
-
-                                if not expected_format or expected_format.startswith(r"\;"):
-                                    if VERSION_REGEX.match(version):
-                                        final_results[software] = [version]
-                                    else:
-                                        final_results[software] = []
-                                elif isinstance(version, str) and r"\;version:" in expected_format:
-                                    final_results[software] = [version]
-                                # Other cases seems to be some kind of false positives
-                            # final_results.update(results)
+                                waiting_version = []
+                                validated = False
+                                for version, js in version_and_js:
+                                    expected_format = data[software]["js"][js]
+                                    if version == "undefined":
+                                        continue
+
+                                    if not expected_format or expected_format.startswith(r"\;confidence:"):
+                                        if VERSION_REGEX.match(version):
+                                            # some false positives here
+                                            final_results[software] = [version]
+                                            break
+                                        if software not in final_results:
+                                            validated = True
+                                            final_results[software] = []
+                                    elif expected_format.startswith(r"\;version:"):
+                                        final_results[software] = [expected_format.split(':')[1]]
+                                    elif isinstance(version, str):
+                                        if r"\;confidence:0" in expected_format:
+                                            waiting_version.append(version)
```

Comment:
what is the story behind those "waiting_version" ?  values we use as a fallback due to lack of confidence ?

Reply:
Sometimes we can detect a version but there is a `confidence:0` in the regex so we are not sure whether the result is relevant or not. Instead of just trash the detected version, we store it in `waiting_version`. If another JS command get a valid result for the current techno, it confirms the presence of the techno. If there is no version already detected we use the one in `waiting_version`.
For instance it is usefull in `underscore.js` and `Lodash` detection where the JS command is the same `_.VERSION`. It allows to detetct only one of them instead of both.

Reply:
Great, thx :)